ufl/exprequals.py

```diff
@@ -48,7 +48,7 @@ def expr_equals(self, other):
                 if s._ufl_typecode_ != o._ufl_typecode_:
                     return False
                 # Skip subtree if objects are the same
-                if s is o or s.ufl_operands is o.ufl_operands:
```

Comment:
The operand comparison here seems to be the main issue. Removing this one makes the test pass.

Anyways, if a pair in the subtree with equal operands gets added to the stack, we can exit early on it with the check above.

Reply:
@schnellerhase could you test this ?:)